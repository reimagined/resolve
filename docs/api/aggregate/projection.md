---
id: projection
title: Projection
---

:::info TypeScript Support

An aggregate projection object has an associated TypeScript type:

- Type Name - `AggregateProjection`
- Containing Package - `@resolve-js/core`

:::

An aggregate projection is an object of the following structure:

```js
const projection = {
  // The *Init* function creates the initial aggregate state object.
  Init: () => initialState,
  // An event handler function is associated with an event type.
  // It receives the aggregate state and an incoming event
  // and returns the updated state.
  [EVENT_TYPE]: (state, event) => {
    ...
    return newState
  }
  [EVENT_TYPE2]: (state, event) => ...
  [EVENT_TYPE3]: (state, event) => ...
  ...
}
```

An event handler implementation receives the following arguments:

| Argument Name | Description                                                   |
| ------------- | ------------------------------------------------------------- |
| state         | The aggregate state that is an object of arbitrary structure. |
| event         | An [event](../event.md) object.                               |

An event handler should return a new state object.
