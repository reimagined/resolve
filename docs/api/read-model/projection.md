---
id: projection
title: Projection
---

:::info TypeScript Support

A read model projection has an associated TypeScript type:

- Type Name - `ReadModel`
- Package - `@resolve-js/core`

:::

A read model projection is an object of the following structure:

```js
const projection = {
  // The *Init* function initializes the read model's persistent store.
  Init: async (store) => { ... }
  // An event handler function is associated with an event type.
  // It receives the read model store and an incoming event
  // and updates the store based on the event's data.
  [EVENT_TYPE1]: async (store, event, context) -> { ... }
  [EVENT_TYPE2]: async (store, event, context) -> { ... }
}
```

An event handler implementation receives the following arguments:

| Argument Name | Description                                                                                              |
| ------------- | -------------------------------------------------------------------------------------------------------- |
| store         | Exposes [API](../read-model/store.md) used to communicate with the read model's persistent data storage. |
| event         | An object that contains the incoming event's data.                                                       |
| context       | An object that contains data and functions related to the current operation.                             |

## context

The `context` argument is an object with the following fields:

| Field Name | Description                                               |
| ---------- | --------------------------------------------------------- |
| encrypt    | The user-defined [encrypt](../../encryption.md) function. |
| decrypt    | The user-defined [decrypt](../../encryption.md) function. |

This object can also contain additional fields added by [middleware](../../middleware.md).
