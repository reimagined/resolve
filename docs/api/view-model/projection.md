---
id: projection
title: Projection
---

:::info TypeScript Support

A view model projection object has an associated TypeScript type:

- Type Name - `ViewModelProjection`
- Package - `@resolve-js/core`

:::

A view model projection is an object of the following structure:

```js
const projection = {
  // The *Init* function creates the view model's initial state object.
  Init: () => initialState,
  // An event handler function is associated with an event type.
  // It receives the view model's state and an incoming event
  // and returns the updated state.
  [EVENT_TYPE]: (state, event) => {
    ...
    return newState
  }
  [EVENT_TYPE2]: (state, event) => ...
  [EVENT_TYPE3]: (state, event) => ...
  ...
}
```

An event handler implementation receives the following arguments:

| Argument Name | Description                                                                                                                                    |
| ------------- | ---------------------------------------------------------------------------------------------------------------------------------------------- |
| state         | The view model's state that is an object of arbitrary structure.                                                                               |
| event         | An [event](../event.md) object.                                                                                                                |
| args          | Arguments attached to the request.                                                                                                             |
| context       | An object that contains functions and data related to the current operation (see the [Event Handler Context](#event-handler-context) section.) |

An event handler should return a new state object.

## Context

A view model event handler context is an object with the following fields:

| Field Name | Description                                               |
| ---------- | --------------------------------------------------------- |
| jwt        | The JSON Web Token attached to the request.               |
| encrypt    | The user-defined [encrypt](../../encryption.md) function. |
| decrypt    | The user-defined [decrypt](../../encryption.md) function. |
