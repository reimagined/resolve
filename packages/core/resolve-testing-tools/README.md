# **resolve-testing-tools**
[![npm version](https://badge.fury.io/js/resolve-testing-tools.svg)](https://badge.fury.io/js/resolve-testing-tools)

This package for testing read-models and sagas.

![Analytics](https://ga-beacon.appspot.com/UA-118635726-1/packages-resolve-testing-tools-readme?pixel)
