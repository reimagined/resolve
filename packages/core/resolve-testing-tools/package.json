{
  "name": "resolve-testing-tools",
  "version": "0.25.2",
  "description": "Tools used to test reSolve applications.",
  "engines": {
    "node": ">=10.9"
  },
  "main": "./lib/index.js",
  "module": "./es/index.js",
  "types": "./types/index.d.ts",
  "bugs": {
    "url": "https://github.com/reimagined/resolve/issues"
  },
  "scripts": {
    "test": "jest --config=../../../jest.config.js --verbose",
    "testw": "jest --config=../../../jest.config.js --watchAll"
  },
  "repository": "https://github.com/reimagined/resolve.git",
  "author": "reimagined team",
  "license": "MIT",
  "keywords": [
    "cqrs",
    "eventsourcing",
    "es",
    "ddd"
  ],
  "homepage": "https://github.com/reimagined/resolve/tree/master/packages/core/resolve-testing-tools#readme",
  "dependencies": {
    "@babel/runtime": "7.9.6",
    "colors": "^1.4.0",
    "diff": "^4.0.2",
    "lodash.isequal": "4.5.0",
    "resolve-command": "0.25.2",
    "resolve-core": "0.25.2",
    "resolve-query": "0.25.2",
    "resolve-readmodel-lite": "0.25.2"
  },
  "devDependencies": {
    "@types/diff": "4.0.2",
    "@types/lodash.isequal": "4.5.5",
    "jest": "26.0.1",
    "sinon": "7.2.2"
  },
  "sourceType": "ts",
  "babelCompile": [
    {
      "moduleType": "cjs",
      "moduleTarget": "client",
      "inputDir": "./src",
      "outDir": "./lib"
    },
    {
      "moduleType": "es",
      "moduleTarget": "client",
      "inputDir": "./src",
      "outDir": "./es"
    }
  ]
}
