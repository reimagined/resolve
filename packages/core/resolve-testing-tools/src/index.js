import createQuery from 'resolve-query'

import as from './as'
import getInitPromise from './get-init-promise'
import givenEvents from './given-events'
import init from './init'
import readModel from './read-model'
import transformEvents from './transform-events'
import initReadModel from './init-read-model'
import initSaga from './init-saga'
import saga from './saga'
import properties from './properties'

export default givenEvents.bind(null, {
  createQuery,
  as,
  getInitPromise,
  init,
  initSaga,
  initReadModel,
  readModel,
  transformEvents,
  saga,
  properties
})

export const RESOLVE_SIDE_EFFECTS_START_TIMESTAMP =
  'RESOLVE_SIDE_EFFECTS_START_TIMESTAMP'
