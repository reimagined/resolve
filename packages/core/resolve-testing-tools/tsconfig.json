{
  "compilerOptions": {
    "target": "es2015",
    "module": "commonjs",
    "moduleResolution": "node",
    "allowJs": true,
    "allowSyntheticDefaultImports": true,
    "esModuleInterop": true,
    "emitDecoratorMetadata": true,
    "experimentalDecorators": true,
    "resolveJsonModule": true,
    "declaration": true,
    "emitDeclarationOnly": true,
    "strict": true,
    "jsx": "react",
    "outDir": "lib",
    "declarationDir": "types",
    "typeRoots": [
      "./node_modules/@types",
      "../../../node_modules/@types"
    ]
  },
  "include": ["src", "test"]
}
