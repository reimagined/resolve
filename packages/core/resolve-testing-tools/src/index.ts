import createQuery from 'resolve-query'
import createCommand from 'resolve-command'
import { SerializableMap, CommandResult } from 'resolve-core'
import as from './as'
import givenEvents from './given-events'
import { execute } from './execute'
import readModel, { ReadModelParams } from './read-model'
import transformEvents from './transform-events'
import { executeReadModel } from './execute-read-model'
import { executeSaga } from './execute-saga'
import { executeCommand } from './execute-command'
import saga, { SagaParams } from './saga'
import properties from './properties'
import getDefaultSecretsManager from './secrets-manager'
import setSecretsManager from './set-secrets-manager'
import { aggregate, BDDAggregate } from './aggregate'
import { command } from './command'

export { BDDAggregate }

export default givenEvents.bind(null, {
  createQuery,
  createCommand,
  as,
  execute,
  executeSaga,
  executeReadModel,
  executeCommand,
  readModel,
  transformEvents,
  saga,
  properties,
  getDefaultSecretsManager,
  setSecretsManager,
  aggregate,
  command,
}) as (events: any[]) => GivenEventsContext

export const RESOLVE_SIDE_EFFECTS_START_TIMESTAMP =
  'RESOLVE_SIDE_EFFECTS_START_TIMESTAMP'

export interface GivenEventsContext {
  all: any
  createQuery: any
  createCommand: any
  as: (jwt: any) => GivenEventsContext
  readModel: (params: ReadModelParams) => GivenEventsContext
  transformEvents: any
  saga: (params: SagaParams) => GivenEventsContext
  properties: (sagaPropertied: any) => GivenEventsContext
  getDefaultSecretsManager: any
  setSecretsManager: any
  aggregate: (
    aggregate: BDDAggregate,
    aggregateId?: string
  ) => GivenEventsContext
  command: (name: string, payload?: SerializableMap) => GivenEventsContext
  shouldProduceEvent: (expectedEvent: CommandResult) => GivenEventsContext
  shouldThrow: (expectedError: any) => GivenEventsContext
}
