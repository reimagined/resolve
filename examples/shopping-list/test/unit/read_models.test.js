import { createReadModel } from 'resolve-testing-tools'

import projection from '../../common/read-models/shopping_lists.projection'
import resolvers from '../../common/read-models/shopping_lists.resolvers'

import {
  SHOPPING_LIST_CREATED,
  SHOPPING_LIST_REMOVED,
  SHOPPING_LIST_RENAMED
} from '../../common/event_types'

describe('read-models', () => {
  describe('ShoppingLists', () => {
    const aggregateId1 = '00000000-0000-0000-0000-000000000000'
    const aggregateId2 = '11111111-1111-1111-1111-111111111111'

    let readModel

    beforeAll(async () => {
      readModel = createReadModel({
        name: 'ShoppingLists',
        projection,
        resolvers
      })
    })

    test('resolver "all" should return an empty array', async () => {
      const shoppingLists = await readModel.resolvers.all()

      expect(shoppingLists).toEqual([])

      expect(shoppingLists).toMatchSnapshot()
    })

    test('projection "SHOPPING_LIST_CREATED" should create a first shopping list', async () => {
      await readModel.applyEvent({
        aggregateId: aggregateId1,
        type: SHOPPING_LIST_CREATED,
        payload: {
          name: 'Products'
        }
      })

      const shoppingLists = await readModel.resolvers.all()

      expect(shoppingLists[0]).toMatchObject({
        id: aggregateId1,
        name: 'Products'
      })

      expect(shoppingLists).toMatchSnapshot()
    })

    test('projection "SHOPPING_LIST_CREATED" should create a second shopping list', async () => {
      await readModel.applyEvent({
        aggregateId: aggregateId2,
        type: SHOPPING_LIST_CREATED,
        payload: {
          name: 'Building materials'
        }
      })

      const shoppingLists = await readModel.resolvers.all()

      expect(shoppingLists[1]).toMatchObject({
        id: aggregateId2,
        name: 'Building materials'
      })

      expect(shoppingLists).toMatchSnapshot()
    })

    test('projection "SHOPPING_LIST_RENAMED" should rename the second shopping list', async () => {
      await readModel.applyEvent({
        aggregateId: aggregateId2,
        type: SHOPPING_LIST_RENAMED,
        payload: {
          name: 'Medicines'
        }
      })

      const shoppingLists = await readModel.resolvers.all()

      expect(shoppingLists[1]).toMatchObject({
        id: aggregateId2,
        name: 'Medicines'
      })

      expect(shoppingLists).toMatchSnapshot()
    })

    test('projection "SHOPPING_LIST_REMOVED" should remove the second shopping list', async () => {
      const beforeRemoveShoppingListsCount = (await readModel.resolvers.all())
        .length

      await readModel.applyEvent({
        aggregateId: aggregateId2,
        type: SHOPPING_LIST_REMOVED
      })

      const shoppingLists = await readModel.resolvers.all()

      expect(beforeRemoveShoppingListsCount - shoppingLists.length).toEqual(1)

      expect(shoppingLists).toMatchSnapshot()
    })
  })
})
