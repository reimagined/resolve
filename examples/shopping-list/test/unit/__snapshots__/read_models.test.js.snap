// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`read-models ShoppingLists projection "SHOPPING_LIST_CREATED" should create a first shopping list 1`] = `
Array [
  Object {
    "createdAt": 1,
    "id": "00000000-0000-0000-0000-000000000000",
    "name": "Products",
  },
]
`;

exports[`read-models ShoppingLists projection "SHOPPING_LIST_CREATED" should create a second shopping list 1`] = `
Array [
  Object {
    "createdAt": 1,
    "id": "00000000-0000-0000-0000-000000000000",
    "name": "Products",
  },
  Object {
    "createdAt": 2,
    "id": "11111111-1111-1111-1111-111111111111",
    "name": "Building materials",
  },
]
`;

exports[`read-models ShoppingLists projection "SHOPPING_LIST_REMOVED" should remove the second shopping list 1`] = `
Array [
  Object {
    "createdAt": 1,
    "id": "00000000-0000-0000-0000-000000000000",
    "name": "Products",
  },
]
`;

exports[`read-models ShoppingLists projection "SHOPPING_LIST_RENAMED" should rename the second shopping list 1`] = `
Array [
  Object {
    "createdAt": 1,
    "id": "00000000-0000-0000-0000-000000000000",
    "name": "Products",
  },
  Object {
    "createdAt": 2,
    "id": "11111111-1111-1111-1111-111111111111",
    "name": "Medicines",
  },
]
`;

exports[`read-models ShoppingLists resolver "all" should return an empty array 1`] = `Array []`;
